src/compiler/checker.ts

```diff
@@ -27181,7 +27181,17 @@ namespace ts {
             // and that call signature is non-generic, return statements are contextually typed by the return type of the signature
             const signature = getContextualSignatureForFunctionLikeDeclaration(functionDecl as FunctionExpression);
             if (signature && !isResolvingReturnTypeOfSignature(signature)) {
-                return getReturnTypeOfSignature(signature);
+                const functionFlags = getFunctionFlags(functionDecl);
+
+                return filterType(getReturnTypeOfSignature(signature), returnType => {
+                    if (functionFlags & FunctionFlags.Generator) {
+                        return checkGeneratorInstantiationAssignabilityToReturnType(returnType, functionFlags, /*errorNode*/ undefined);
+                    }
+                    if (functionFlags & FunctionFlags.Async) {
+                        return !!getAwaitedTypeOfPromise(returnType);
+                    }

```

Comment:
This block supersedes [my other attempt from a couple of months back](https://github.com/microsoft/TypeScript/pull/47683/files#diff-d9ab6589e714c71e657f601cf30ff51dfc607fc98419bf72e04f6b0fa92cc4b8) at fixing #47682 . I think that the fix here is better but perhaps that older one might be still interesting for reviewers